### ISSUE TITLE
Sphinx build hanging on file-formats.md

### ISSUE THREAD

**Issue by kinow** [common-workflow-language/user_guide#268]
I have this issue on my environment, but GH Actions is clearly not having the same problem. I'll document it here in case others have a similar issue, or if someone would like to dig into it to understand what's happening.

1. I've synced my `main` branch, 

```bash
kinow@ranma:~/Development/python/workspace/user_guide$ git log -n 1
commit 50d824f23edf76012d7d4002681e4f9f24927b4b (HEAD -> main, upstream/main, upstream/HEAD)
Author: Bruno P. Kinoshita <kinow@users.noreply.github.com>
Date:   Mon Oct 10 22:58:10 2022 +1300

    Add section about exclusive params with expressions, and emphasize command output lines (#247)
    
    * Highlight lines of interest in the command outputs
    * Rename "Advanced Inputs" to "Inclusive and Exclusive Inputs"
    * Add text about the exclusive input parameters and expressions
    * explain why itemC was chosen when both are provided
    
    Co-authored-by: Michael R. Crusoe <michael.crusoe@gmail.com>
```

2. I've created a new environment with venv (used Mamba to start a dev env for CWL projects, but it's a fresh new env):

```bash
$ mamba create -n dev python==3.10 --quiet --yes

                  __    __    __    __
                 /  \  /  \  /  \  /  \
                /    \/    \/    \/    \
███████████████/  /██/  /██/  /██/  /████████████████████████
              /  / \   / \   / \   / \  \____
             /  /   \_/   \_/   \_/   \    o \__,
            / _/                       \_____/  `
            |/
        ███╗   ███╗ █████╗ ███╗   ███╗██████╗  █████╗
        ████╗ ████║██╔══██╗████╗ ████║██╔══██╗██╔══██╗
        ██╔████╔██║███████║██╔████╔██║██████╔╝███████║
        ██║╚██╔╝██║██╔══██║██║╚██╔╝██║██╔══██╗██╔══██║
        ██║ ╚═╝ ██║██║  ██║██║ ╚═╝ ██║██████╔╝██║  ██║
        ╚═╝     ╚═╝╚═╝  ╚═╝╚═╝     ╚═╝╚═════╝ ╚═╝  ╚═╝

        mamba (0.25.0) supported by @QuantStack

        GitHub:  https://github.com/mamba-org/mamba
        Twitter: https://twitter.com/QuantStack

█████████████████████████████████████████████████████████████

  Package               Version  Build               Channel                    Size
──────────────────────────────────────────────────────────────────────────────────────
  Install:
──────────────────────────────────────────────────────────────────────────────────────

  + _libgcc_mutex           0.1  conda_forge         conda-forge/linux-64     Cached
  + _openmp_mutex           4.5  2_gnu               conda-forge/linux-64     Cached
  + bzip2                 1.0.8  h7f98852_4          conda-forge/linux-64     Cached
  + ca-certificates   2022.9.24  ha878542_0          conda-forge/linux-64     Cached
  + ld_impl_linux-64     2.36.1  hea4e1c9_2          conda-forge/linux-64     Cached
  + libffi                3.4.2  h7f98852_5          conda-forge/linux-64     Cached
  + libgcc-ng            12.1.0  h8d9b700_16         conda-forge/linux-64     Cached
  + libgomp              12.1.0  h8d9b700_16         conda-forge/linux-64     Cached
  + libnsl                2.0.0  h7f98852_0          conda-forge/linux-64     Cached
  + libsqlite            3.39.4  h753d276_0          conda-forge/linux-64     Cached
  + libuuid              2.32.1  h7f98852_1000       conda-forge/linux-64     Cached
  + libzlib              1.2.12  h166bdaf_4          conda-forge/linux-64     Cached
  + ncurses                 6.3  h27087fc_1          conda-forge/linux-64     Cached
  + openssl               3.0.5  h166bdaf_2          conda-forge/linux-64     Cached
  + pip                  22.2.2  pyhd8ed1ab_0        conda-forge/noarch       Cached
  + python               3.10.0  h543edf9_3_cpython  conda-forge/linux-64     Cached
  + readline              8.1.2  h0f457ee_0          conda-forge/linux-64     Cached
  + setuptools           65.4.1  pyhd8ed1ab_0        conda-forge/noarch       Cached
  + sqlite               3.39.4  h4ff8645_0          conda-forge/linux-64     Cached
  + tk                   8.6.12  h27826a3_0          conda-forge/linux-64     Cached
  + tzdata                2022d  h191b570_0          conda-forge/noarch       Cached
  + wheel                0.37.1  pyhd8ed1ab_0        conda-forge/noarch       Cached
  + xz                    5.2.6  h166bdaf_0          conda-forge/linux-64     Cached

  Summary:

  Install: 23 packages

  Total download: 0 B

──────────────────────────────────────────────────────────────────────────────────────

Preparing transaction: ...working... done
Verifying transaction: ...working... done
Executing transaction: ...working... done
$ mamba activate dev
(dev) $ python -m venv venv
(dev) $ source venv/bin/activate
(dev) $ pip install -U pip
Requirement already satisfied: pip in ./venv/lib/python3.10/site-packages (21.2.3)
Collecting pip
  Using cached pip-22.2.2-py3-none-any.whl (2.0 MB)
Installing collected packages: pip
  Attempting uninstall: pip
    Found existing installation: pip 21.2.3
    Uninstalling pip-21.2.3:
      Successfully uninstalled pip-21.2.3
Successfully installed pip-22.2.2
(dev) $ pip install -e .[all]
# (...)
  Running setup.py install for livereload ... done
  Running setup.py install for cwl-runner ... done
Successfully installed CacheControl-0.12.11 MarkupSafe-2.1.1 alabaster-0.7.12 argcomplete-2.0.0 babel-2.10.3 bagit-1.8.1 beautifulsoup4-4.11.1 certifi-2022.9.24 charset-normalizer-2.1.1 colorama-0.4.5 coloredlogs-15.0.1 cwl-runner-1.0 cwl-upgrader-1.2.4 cwl-user-guide-0.1 cwl-utils-0.20 cwltest-2.2.20220521103021 cwltool-3.1.20221008225030 defusedxml-0.7.1 docutils-0.19 humanfriendly-10.0 idna-3.4 imagesize-1.4.1 isodate-0.6.1 jinja2-3.1.2 junit-xml-1.9 livereload-2.6.3 lockfile-0.12.2 lxml-4.9.1 markdown-it-py-2.1.0 mdit-py-plugins-0.3.1 mdurl-0.1.2 mistune-0.8.4 msgpack-1.0.4 mypy-extensions-0.4.3 myst-parser-0.18.1 networkx-2.8.7 packaging-21.3 prov-1.5.1 psutil-5.9.2 pydata-sphinx-theme-0.11.0 pydot-1.4.2 pygments-2.13.0 pyparsing-3.0.9 python-dateutil-2.8.2 pytz-2022.4 pyyaml-6.0 rdflib-6.2.0 requests-2.28.1 ruamel.yaml-0.17.21 ruamel.yaml.clib-0.2.6 schema-salad-8.3.20220916115321 shellescape-3.8.1 six-1.16.0 snowballstemmer-2.2.0 soupsieve-2.3.2.post1 sphinx-5.2.3 sphinx-autobuild-2021.3.14 sphinx-reredirects-0.1.1 sphinxcontrib-applehelp-1.0.2 sphinxcontrib-devhelp-1.0.2 sphinxcontrib-htmlhelp-2.0.0 sphinxcontrib-jsmath-1.0.1 sphinxcontrib-qthelp-1.0.3 sphinxcontrib-runcmd-0.2.0 sphinxcontrib-serializinghtml-1.1.5 tornado-6.2 typing-extensions-4.4.0 udocker-1.3.4 urllib3-1.26.12
```

3. Now with all the dependencies installed in my environment, I tried the following:

```bash
(dev) $ rm -rf _build/ && make build
Running Sphinx v5.2.3
^C
Interrupted!
make: *** [Makefile:50: build] Interrupt

(venv) (dev) kinow@ranma:~/Development/python/workspace/user_guide$ rm -rf _build/ && make html
Running Sphinx v5.2.3
making output directory... done
myst v0.18.1: MdParserConfig(commonmark_only=False, gfm_only=False, enable_extensions=['colon_fence', 'deflist', 'substitution', 'replacements'], disable_syntax=[], all_links_external=False, url_schemes=('http', 'https', 'mailto', 'ftp'), ref_domains=None, highlight_code_blocks=True, number_code_blocks=[], title_to_header=False, heading_anchors=4, heading_slug_func=None, footnote_transition=True, words_per_minute=200, sub_delimiters=('{', '}'), linkify_fuzzy_links=True, dmath_allow_labels=True, dmath_allow_space=True, dmath_allow_digits=True, dmath_double_inline=False, update_mathjax=True, mathjax_classes='tex2jax_process|mathjax_process|math|output_area')
building [mo]: targets for 0 po files that are out of date
building [html]: targets for 33 source files that are out of date
updating environment: [new config] 33 added, 0 changed, 0 removed
reading sources... [ 57%] topics/file-formats 
```

4. That was at 14:52. Still at 14:52 it showed the message “reading sources... [57&] topics/file-formats”, as in the screenshot below.

![image](https://user-images.githubusercontent.com/304786/194979569-5ba2b199-7006-4a05-9534-12cb4f85cd36.png)

5. At 14:55, I took this screenshot of `htop` (configured to Display tree, then scrolling down to locate the `make html` command entry):

![image](https://user-images.githubusercontent.com/304786/194979810-9a624db8-fcc9-45d7-918a-03b03015d3ba.png)

6. I stopped the command at 15:00.
7. This also hangs:

```bash
(dev) $  cd src/_includes/cwl/file-formats/
(dev) $ cwltool --debug metadata_example.cwl sample.yml 
INFO /home/kinow/Development/python/workspace/user_guide/venv/bin/cwltool 3.1.20221008225030
INFO Resolved 'metadata_example.cwl' to 'file:///home/kinow/Development/python/workspace/user_guide/src/_includes/cwl/file-formats/metadata_example.cwl'
```

I haven't had time to dig really deep into the issue, but this makes the workflow run in about 1 second, so it might be something with how `$schemas` is handled, and maybe my `cwltool` version (above), or my environment.

```diff
$ git diff
diff --git a/src/_includes/cwl/file-formats/metadata_example.cwl b/src/_includes/cwl/file-formats/metadata_example.cwl
index 9485522..f4b9482 100644
--- a/src/_includes/cwl/file-formats/metadata_example.cwl
+++ b/src/_includes/cwl/file-formats/metadata_example.cwl
@@ -24,5 +24,5 @@ outputs:
 
 $namespaces:
   edam: http://edamontology.org/
-$schemas:
-  - http://edamontology.org/EDAM_1.18.owl
+#$schemas:
+#  - http://edamontology.org/EDAM_1.18.owl
```

Thanks!

**Comment by mr-c** [comment]
Hey @kinow ; I'm confused as why you use mamba, when the first step is to make & activate a virtualenv. Maybe that is causing the issue?

**Comment by kinow** [comment]
Hi @mr-c! I never used my Ubuntu Python. I used miniconda before, but replaced it recently by mamba (not minimamba). But good point! Let me test with Ubuntu's python3 creating a venv… just a sec.

**Comment by kinow** [comment]
Tested with Ubuntu LTS' `python3 -m venv venv` after removing my old `venv` created with mamba.

```
(venv) kinow@ranma:~/Development/python/workspace/user_guide$ python
Python 3.10.6 (main, Aug 10 2022, 11:40:04) [GCC 11.3.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> 
(venv) kinow@ranma:~/Development/python/workspace/user_guide$ cwltool --version
/home/kinow/Development/python/workspace/user_guide/venv/bin/cwltool 3.1.20221008225030
(venv) kinow@ranma:~/Development/python/workspace/user_guide$ git log -n 1
commit bb948f06736d0583dfef81cf185a857ea97d3587 (HEAD -> main, upstream/main, upstream/HEAD)
Author: Michael R. Crusoe <michael.crusoe@gmail.com>
Date:   Wed Oct 12 12:00:54 2022 +0200

    install: upgrade pip/setuptools/wheel first
```

:+1: 
![image](https://user-images.githubusercontent.com/304786/195319452-3f345201-d8ff-4351-a625-1e2ae8f0751d.png)

4 minutes stuck and counting.

**Comment by kinow** [comment]
Still stuck. Doesn't seem to be due to mamba. But it was a good idea. I'm watching a webinar, but waiting for the parts that interest me (still 1 hour away :sob:) so I think I can try to debug this workflow on my environment directly with cwltool. I'll report here or create an issue there if I find anything :+1:

**Comment by mr-c** [comment]
Thanks for testing; I also tried to reproduce on my side without `nodejs` installed, but it still worked here

**Comment by kinow** [comment]
I didn't think about that, another good shot, thanks! I have the debugger open now. Already found that it hangs in the `resolve_and_validate_document`. So running `cwltool --validate metadata_example.cwl sample.yml` is enough to reproduce it for me.

**Comment by kinow** [comment]
Ah! Solved it! Doesn't require any code change.

1. I started debugging it in `cwltool`, with a command running the exact same thing as the command above, but without the `--validate`.
2. I followed the code through `resolve_and_validate_document`, then to `schema_salad`'s `fetcher.py`. And then to `requests`.
3. In `requests` kept following the code until it reached `cachecontrol`.
4. Finally, I took note of the headers, and tried to execute the same thing in `curl`. I got a 304. So I knew the cache would be used.

```bash
kinow@ranma:/tmp$ curl -H "User-Agent: python-requests/2.28.1" -H "Accept-Encoding: gzip, deflate" -H "Accept: */*" -H "Connection: keep-alive" -H 'If-None-Match: "29ff38-553a265b94000"' -I http://edamontology.org/EDAM_1.18.owl 
HTTP/1.1 304 Not Modified
Date: Wed, 12 Oct 2022 11:45:16 GMT
Server: Apache/2.4.6 (CentOS) OpenSSL/1.0.2k-fips mod_fcgid/2.3.9 PHP/5.6.40 mod_wsgi/3.4 Python/2.7.5 mod_perl/2.0.11 Perl/v5.16.3
Connection: Keep-Alive
Keep-Alive: timeout=5, max=100
ETag: "29ff38-553a265b94000"
```

5. Followed the rest of the code, and found a local directory `~/.cache/salad/` where `cachecontrol` stores the responses of `requests`.
6. Deleted the whole directory.
7. Everything works.

![Screenshot from 2022-10-13 00-11-40](https://user-images.githubusercontent.com/304786/195340070-c38571ce-9fcc-47d8-9780-fd287b4734a0.png)

![Screenshot from 2022-10-13 00-57-22](https://user-images.githubusercontent.com/304786/195340094-693580f9-0794-4aa1-b0c3-6cf1e04a4936.png)

@mr-c I suspect you were probably correct. It might have been caused by mamba, and cachecontrol getting confused when trying to load the already saved cache files.

Tested `make html` and now the build succeeded :+1: Lesson learned: if nothing works, delete the `.cache/salad` directory.

Cheers
Bruno
